src/stream.rs

```diff
@@ -0,0 +1,263 @@
+// Copyright 2015-2016 Mozilla Foundation. See the COPYRIGHT
+// file at the top-level directory of this distribution.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::fmt;
+use std::io;
+use std::mem;
+use std::ops::Range;
+use std::str;
+use super::{Decoder, Encoder};
+
+pub use std::io::Write as WriteBytes;
+pub use std::io::Read as ReadBytes;
+pub use std::fmt::Write as WriteUnicode;
+
+pub trait ReadUnicode {
+    fn read(&mut self, dst: &mut str) -> io::Result<usize>;
+}
+
+impl Decoder {
+    pub fn wrap_read<S, B>(self, stream: S, buffer: B) -> ReadDecoder<S, B>
+    where S: ReadBytes, B: AsMut<[u8]> {
+        ReadDecoder {
+            decoder: self,
+            stream: stream,
+            reached_stream_eof: false,
+            buffer: buffer,
+            unused_buffer_slice: 0..0,
+        }
+    }
+
+    pub fn wrap_write<S, B>(self, stream: S, buffer: B) -> WriteDecoder<S, B>
+    where S: WriteUnicode, B: AsMut<[u8]> {
+        WriteDecoder {
+            decoder: self,
+            stream: stream,
+            buffer: buffer,
+        }
+    }
+}
+
+impl Encoder {
+    pub fn wrap_read<S, B>(self, stream: S, buffer: B) -> ReadEncoder<S, B>
+    where S: ReadUnicode, B: AsMut<str> {
+        ReadEncoder {
+            encoder: self,
+            stream: stream,
+            reached_stream_eof: false,
+            buffer: buffer,
+            unused_buffer_slice: 0..0,
+        }
+    }
+
+    pub fn wrap_write<S, B>(self, stream: S, buffer: B) -> WriteEncoder<S, B>
+    where S: WriteBytes, B: AsMut<[u8]> {
+        WriteEncoder {
+            encoder: self,
+            stream: stream,
+            buffer: buffer,
+        }
+    }
+}
+
+impl<'a, S: ReadUnicode> ReadUnicode for &'a mut S {
+    fn read(&mut self, dst: &mut str) -> io::Result<usize> {
+        (**self).read(dst)
+    }
+}
+
+/// Similar to `std::io::Read for &'a [u8]`.
+impl<'a> ReadUnicode for &'a str {
+    fn read(&mut self, dst: &mut str) -> io::Result<usize> {
+        // Unsafe: copy_maximum_utf8_prefix carefully preserves UTF-8 wel-formedness.
+        let dst_bytes = unsafe { mem::transmute::<&mut str, &mut [u8]>(dst) };
+        let (bytes_copied, remaining_self) = copy_maximum_utf8_prefix(*self, dst_bytes);
+        *self = remaining_self;
+        Ok(bytes_copied)
+    }
+}
+
+fn copy_maximum_utf8_prefix<'a>(src: &'a str, dst: &mut [u8]) -> (usize, &'a str) {
+    let src_bytes = src.as_bytes();
+    let bytes_to_copy = if dst.len() >= src.len() {
+        src.len()
+    } else {
+        let mut i = dst.len() - 1;
```

Comment:
Why `- 1`? We know `dst.len() < src.len()`, so `src[dst.len()]` won't panic. If `src[dst.len()]` is a UTF-8 lead byte, things should be fine.